### PR TITLE
Reorganize RenderingDevice barriers

### PR DESCRIPTION
**Pull request by reduz**
-Removed sync to draw, now everything syncs to draw by default.
-Fixed many validation layer errors.
-Added support for VkImageViewUsageCreateInfo to fix validation layer warnings.
-Texture, buffer, raster and compute functions now all allow spcifying which barriers will be used.

### CODE REVIEW

File: servers/rendering/rendering_device.cpp [hunk 1]
```diff
@@ -249,17 +249,17 @@ void RenderingDevice::_bind_methods() {
...
-	ClassDB::bind_method(D_METHOD("texture_clear", "texture", "color", "base_mipmap", "mipmap_count", "base_layer", "layer_count", "sync_with_draw"), &RenderingDevice::texture_clear, DEFVAL(false));
-	ClassDB::bind_method(D_METHOD("texture_resolve_multisample", "from_texture", "to_texture", "sync_with_draw"), &RenderingDevice::texture_resolve_multisample, DEFVAL(false));
+	ClassDB::bind_method(D_METHOD("texture_copy", "from_texture", "to_texture", "from_pos", "to_pos", "size", "src_mipmap", "dst_mipmap", "src_layer", "dst_layer", "post_barrier"), &RenderingDevice::texture_copy, DEFVAL(BARRIER_MASK_ALL));
+	ClassDB::bind_method(D_METHOD("texture_clear", "texture", "color", "base_mipmap", "mipmap_count", "base_layer", "layer_count", "post_barrier"), &RenderingDevice::texture_clear, DEFVAL(BARRIER_MASK_ALL));
+	ClassDB::bind_method(D_METHOD("texture_resolve_multisample", "from_texture", "to_texture", "post_barrier"), &RenderingDevice::texture_resolve_multisample, DEFVAL(false));
```
Comment (clayjohn) [inline servers/rendering/rendering_device.cpp:262]: The DEFVAL here needs to update

**Review by clayjohn — CHANGES_REQUESTED** [review]

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 2]
```diff
@@ -41,28 +41,60 @@
...
+			if (p_post_barrier & BARRIER_MASK_RASTER) {
+				access_mask |= VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT;
+				stage_mask |= VK_PIPELINE_STAGE_VERTEX_SHADER_BIT | VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT;
+			}
+			if (p_post_barrier & BARRIER_MASK_COMPUTE) {
+				access_mask |= VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_SHADER_WRITE_BIT;
+				stage_mask |= VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT;
+			}
+		}
 	} else if (index_buffer_owner.owns(p_buffer)) {
```
[the remark this answers is not in the document]
Comment (reduz) [reply drivers/vulkan/rendering_device_vulkan.cpp:61]: for now you cant really write to it in compute. Could add the feature eventually then barrier will have to be added

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 3]
```diff
@@ -41,28 +41,60 @@
 //#define FORCE_FULL_BARRIER
 
 // Get the Vulkan object information and possible stage access types (bitwise OR'd with incoming values)
-RenderingDeviceVulkan::Buffer *RenderingDeviceVulkan::_get_buffer_from_owner(RID p_buffer, VkPipelineStageFlags &stage_mask, VkAccessFlags &access_mask) {
+RenderingDeviceVulkan::Buffer *RenderingDeviceVulkan::_get_buffer_from_owner(RID p_buffer, VkPipelineStageFlags &stage_mask, VkAccessFlags &access_mask, uint32_t p_post_barrier) {
```
Comment (JFonS) [inline drivers/vulkan/rendering_device_vulkan.cpp:44]: I would add the `r_` prefix to the last two arguments.

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 4]
```diff
@@ -5038,70 +5182,80 @@ bool RenderingDeviceVulkan::uniform_set_is_valid(RID p_uniform_set) {
...
 
-Error RenderingDeviceVulkan::buffer_update(RID p_buffer, uint32_t p_offset, uint32_t p_size, const void *p_data, bool p_sync_with_draw) {
+Error RenderingDeviceVulkan::buffer_update(RID p_buffer, uint32_t p_offset, uint32_t p_size, const void *p_data, uint32_t p_post_barrier) {
 	_THREAD_SAFE_METHOD_
 
-	ERR_FAIL_COND_V_MSG(draw_list && p_sync_with_draw, ERR_INVALID_PARAMETER,
+	ERR_FAIL_COND_V_MSG(draw_list, ERR_INVALID_PARAMETER,
 			"Updating buffers in 'sync to draw' mode is forbidden during creation of a draw list");
-	ERR_FAIL_COND_V_MSG(compute_list && p_sync_with_draw, ERR_INVALID_PARAMETER,
+	ERR_FAIL_COND_V_MSG(compute_list, ERR_INVALID_PARAMETER,
```
Comment (JFonS) [inline drivers/vulkan/rendering_device_vulkan.cpp:5190]: The error messages can probably be updated.

File: servers/rendering/rendering_device.h [hunk 5]
```diff
@@ -649,8 +660,8 @@ class RenderingDevice : public Object {
 	virtual RID uniform_set_create(const Vector<Uniform> &p_uniforms, RID p_shader, uint32_t p_shader_set) = 0;
 	virtual bool uniform_set_is_valid(RID p_uniform_set) = 0;
 
-	virtual Error buffer_update(RID p_buffer, uint32_t p_offset, uint32_t p_size, const void *p_data, bool p_sync_with_draw = false) = 0; //this function can be used from any thread and it takes effect at the beginning of the frame, unless sync with draw is used, which is used to mix updates with draw calls
-	virtual Error buffer_clear(RID p_buffer, uint32_t p_offset, uint32_t p_size, bool p_sync_with_draw = false) = 0;
+	virtual Error buffer_update(RID p_buffer, uint32_t p_offset, uint32_t p_size, const void *p_data, uint32_t p_post_barrier = BARRIER_MASK_ALL) = 0; //this function can be used from any thread and it takes effect at the beginning of the frame, unless sync with draw is used, which is used to mix updates with draw calls
```
Comment (JFonS) [inline servers/rendering/rendering_device.h:663]: The comment needs to be updated or removed.

**Review by JFonS — APPROVED** [review]

**Review by clayjohn — APPROVED** [review]

**Pull request merged**